src/choi_operators.jl

```diff
@@ -0,0 +1,141 @@
+and a ``0`` everywhere else. The input ``\\Lambda`` is the output dimension.
+
+The input function `𝒩` is called as `𝒩(X)` for arbitrary input `X`.
+Channel functions with multiple parameters can be considered by declaring
+a function
+`𝒩_xy(ρ) = 𝒩(ρ,x,y)` for fixed `(x,y)` and then call, `choi(𝒩_xy, dim_in, dim_out)`.
+"""
+function choi_matrix(𝒩 :: Function, dims :: Vector{Int}) :: Matrix{ComplexF64}
+    dim_in, dim_out = dims
+
+    eab_matrix = zeros(ComplexF64, dim_in , dim_in)
+    Λ = zeros(ComplexF64, dim_in * dim_out, dim_in * dim_out)
+    for i in 1:dim_in
+        row_ids = ((i-1) * dim_out + 1):((i-1) * dim_out + dim_out)
+        for j in 1:dim_out
+            col_ids = ((j-1) * dim_out + 1):((j-1) * dim_out + dim_out)
+
+            eab_matrix[i,j] = 1
+            Λ[row_ids,col_ids] += 𝒩(eab_matrix)
+            eab_matrix[i,j] = 0
+        end
+    end
+
+    return Λ
+end
+
+"""
+    ChoiOp( Λ :: AbstractMatrix, dims :: Vector{Int} ) :: ChoiOp{<:Number}
+    ChoiOp( 𝒩 :: Function, dims :: Vector{Int} ) :: ChoiOp{ComplexF64}
+
+Constructs the Choi operator representation of a quantum channel.
+If either a function `𝒩` or set of kraus operators is provided as input, the
+Choi matrix is constructed with the [`choi_matrix`](@ref) method.
+
+The `ChoiOp` type contains the fields:
+* `M :: Matrix{<:Number}` - The Choi matrix.
+* `dims :: Vector{Int}` - The channel's input/output dimension `[dim_in, dim_out]`.
+
+A `DomainError` is thrown if [`is_choi_matrix`](@ref) returns `false`.
+"""
+struct ChoiOp{T<:Number} <: Operator{T}
+    M :: Matrix{T}
+    dims :: Vector{Int}
+    atol :: Float64
+    ChoiOp(
+        Λ :: AbstractMatrix{T},
+        dims :: Vector{Int};
+        atol=ATOL :: Float64
+    ) where T <: Number = is_choi_matrix(Λ, dims, atol=atol) ? new{T}(Λ, dims, atol) : throw(
+        DomainError(Λ, "The Choi operator is not a valid quantum channel.")
+    )
+    ChoiOp(
+        𝒩 :: Function,
+        dims :: Vector{Int};
+        atol=ATOL :: Float64
+    ) = ChoiOp( choi_matrix(𝒩, dims), dims, atol=atol)
+end
+
+# print out matrix forms when Choi types are displayed
+function show(io::IO, mime::MIME{Symbol("text/plain")}, Λ :: ChoiOp)
+    summary(io, Λ)
+    print(io, "\ndims : ", Λ.dims)
+    print(io, "\natol : ", Λ.atol)
+    print(io, "\nM : ")
+    show(io, mime, Λ.M)
+end
+
+"""
+    choi_evolve(Λ :: Matrix{<:Number}, ρ :: Matrix{<:Number}) :: Matrix
+    choi_evolve(
+        Λ :: Matrix{<:Number}, ρ :: Matrix{<:Number}, dims :: Vector{Int}
+    ) :: Matrix
+
+Applies the Choi operator `Λ` to the density operator `ρ`. The output of this
+quantum channel evaluated as,
+
+```math
+    \\rho'_B = \\text{Tr}_A[\\Lambda_{AB}(\\rho_A^{T}\\otimes \\mathbb{I}_B],
+```
+
+where the [`partial_trace`](@ref) is take with respect to the input system.
+"""
+function choi_evolve(Λ :: Matrix{<:Number}, ρ :: Matrix{<:Number}) :: Matrix
+    dim_in = size(ρ, 1)
+    dim_out = size(Λ, 1) ÷ dim_in
+
```

Comment:
I'm actually inclined to keep and here's why. I've written an `evolve(Λ :: ChoiOp, ρ :: State)`  for `ChoiOp` types. However, some of the criticism I've faced with QBase.jl is that it adds too much type abstraction i.e. some people want quick and dirty matrix operations over verified quantum operator types.

I somewhat agree because it is annoying to be forced into casting a matrix as a ChoiOp before running a method. Likewise it is annoying for quantum objects to be verified with time consuming operations like checking positive semi-definite-ness. 

In response to these criticisms, I want to make quantum operations available that do not use the type structure or verification. This approach is the same as giving people guns, and if they shoot themselves in the foot, it's their own damn fault. Some people prefer guns, others, the safety net of type structures. 

In response to your comment, I've added a warning to the docs about potential errors related to improper dimension.